### PR TITLE
Fix the concurrency problem when AbstractCacheFactory creates Cache

### PR DESCRIPTION
**Pull request by xiaoheng1**
## What is the purpose of the change

Fix the concurrency problem when AbstractCacheFactory creates Cache

see more detail from #7674 

## Verifying this change

- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java [hunk 1]
```diff
@@ -42,11 +43,14 @@
     /**
      * This is used to store factory level-1 cached data.
      */
-    private final ConcurrentMap<String, Cache> caches = new ConcurrentHashMap<String, Cache>();
+    private final static ConcurrentMap<String, Cache> caches = new ConcurrentHashMap<String, Cache>();
```
Comment (horizonzy) [inline dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:46]: 没必要用static,不同种类的Cache用自己的caches

Comment (xiaoheng1) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:46]: In my opinion, AbstractCacheFactory.caches should store jcache, lfu, lru cache, etc. Therefore, I changed caches to static, and let caches be shared by all other cache implementation classes. such as AbstractRegsitryFactory.

Comment (horizonzy) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:46]: I use both lra and lfu, if it share cache, how did eliminate the data.

Comment (xiaoheng1) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java]: done

File: dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java [hunk 2]
```diff
@@ -23,6 +23,7 @@
 
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.ReentrantLock;
```
Comment (horizonzy) [inline dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:26]: unless import.

**Review by horizonzy — APPROVED** [review]
LGTM.

File: dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java [hunk 3]
```diff
@@ -55,15 +58,29 @@ public Cache getCache(URL url, Invocation invocation) {
         url = url.addParameter(METHOD_KEY, invocation.getMethodName());
```
Comment (pinxiong) [inline dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: You'd better not overwrite the value of `url`

Comment (xiaoheng1) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: I did not modify this piece of code. Can you describe your opinion in detail?

Comment (pinxiong) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: The `url` 's value is changed
```java
public Cache getCache(URL url, Invocation invocation) {
    url = url.addParameter(METHOD_KEY, invocation.getMethodName());
    String key = url.toFullString();
    Cache cache = caches.get(key);
}
```

The `url` 's value is unchanged
```java
public Cache getCache(URL url, Invocation invocation) {
    String key = url.addParameter(METHOD_KEY, invocation.getMethodName()).toFullString();
    Cache cache = caches.get(key);
}
```

I know you don't modify this piece of code, but I think it's unreasonable to change the value of an input parameter if the caller don't know that.

Also, I know that if you do it the way I said, you might need to change a lot of code (All the places where `METHOD_KEY` is used may need to be modified)

Comment (xiaoheng1) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: The URL is designed to be immutable. After adding parameters, if the url is not accepted, the new url cannot be obtained. If the METHOD_KEY parameter is added elsewhere, this parameter may not be needed in that place. Do you have any good suggestions?

Comment (pinxiong) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: This is indeed a complicated problem. You will change a lot of code if you want to solve this problem completely.

You can refer #7641

For this issue, your solution may be the simplest and least risky

Comment (xiaoheng1) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: Good suggestion. 

I think we should fix this problem first. The problem of url reconstruction will be dealt with later. What do you think so?

Comment (pinxiong) [reply dubbo-filter/dubbo-filter-cache/src/main/java/org/apache/dubbo/cache/support/AbstractCacheFactory.java:58]: Yes, you're right!

**Review by AlbumenJ — APPROVED** [review]

**Pull request merged**